pulp_container/app/access_policy.py

```diff
@@ -6,46 +6,34 @@
 from pulp_container.app import models
 
 
-class NamespacePermissionsChecker:
+class NamespaceAccessPolicyMixin:
     """
-    A class that contains a function which checks permissions required for modifying namespaces.
+    Access policy mixin for ContainerDistributionViewSet and ContainerPushRepositoryViewSet which
+    handles namespace permissions.
     """
 
-    @staticmethod
-    def has_permissions(namespace, user, permission):
+    def has_namespace_obj_perms(self, request, view, action, permission):
         """
-        Check whether a user have permissions to manage the passed namespace.
+        Check if a user has object-level perms on the namespace associated with the distribution
+        or repository.
         """
-
-        try:
-            namespace = models.ContainerNamespace.objects.get(name=namespace)
-        except models.ContainerNamespace.DoesNotExist:
-            # check model permissions for namespace creation
-            return user.has_perm("container.add_containernamespace")
+        obj = view.get_object()
+        if type(obj) == models.ContainerDistribution:
+            namespace = obj.namespace
         else:
-            # existing namespace
-            return user.has_perm(permission) or user.has_perm(permission, namespace)
-
-
-class DistributionAccessPolicyMixin:
-    """
-    Access policy mixin for DistributionViewSet which handles namespace permissions.
-    """
-
-    def has_manage_namespace_dist_perms(self, request, view, action, permission):
-        """
-        Check whether a user can create a namespace or it can manage distributions
-        for existing namespace.
-        """
-        namespace = request.data["base_path"].split("/")[0]
-        return NamespacePermissionsChecker.has_permissions(namespace, request.user, permission)
+            distribution = models.ContainerDistribution.objects.get(repository=obj)
```

Comment:
Theoretically a repository can be distributed by mutiple distributions.
Another distribution can be created via pulp API and repo manually assigned to the distribution.
I suggest:
```
dists_qs = models.ContainerDistribution.objects.filter(repository=obj)
for dist in dists_qs:
    if request.user.has_perm(permission, dist.namespace):
        return True
```

Reply:
That makes sense. I'll make the change.